### DISCUSSION TITLE
[Proposal] Support emoji for variable name 🍍🐶🐮🚀

### DISCUSSION THREAD

**Discussion by NicolasDorier** [dotnet/csharplang#3490] [General]
To my disappointment, variable names does not support all unicode characters

# How to reproduce
```csharp
public enum ThreeStateBool
{
    🍍,
    Yes,
    No
}
```

# Expected

The code build and I can use fruits to better communicate code requirements via variable names.

# Actual

A build error
```
Error	CS1056	Unexpected character '�'	
```

I am using C# 8.0.

**Comment by ufcpp** [comment]
see https://github.com/dotnet/csharplang/issues/1742#issuecomment-633200128

However, 🍍 can not be an identifier because its category is Other Symbol.

**Comment by NicolasDorier** [comment]
Ah what a disappointment :( closing as dup then.

**Comment by NicolasDorier** [comment]
Reopening as this is a different request on valid unicode for variable name.

This has been implemented in g++ 10. Fixing issue https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67224

g++ 10.1.1 +  binutils 2.34.0.20200522)

![image](https://user-images.githubusercontent.com/3020646/82916396-f31a4d80-9fac-11ea-8c0f-3ca98bd21b82.png)

**Comment by HaloFour** [comment]
I, for one, would gladly welcome any language changes that would include more 🥓.

Swift is another language that allows for characters such as emoji in identifiers:

```swift
let 🐶🐮 = "dogcow"
let ＃🚀 = "Hashrocket"
NSLog(＃🚀)
```

**Comment by svick** [comment]
Note that the C# spec (mostly) attempts to follow [Unicode Standard Annex #31: Unicode Identifier and Pattern Syntax](https://unicode.org/reports/tr31/), which does not allow emoji.

**Comment by Unknown6656** [comment]
> Note that the C# spec (mostly) attempts to follow [Unicode Standard Annex #31: Unicode Identifier and Pattern Syntax](https://unicode.org/reports/tr31/), which does not allow emoji.

Thank heavens for that!
I already see some autonomous vehicles crashing into walls due to programming bugs in the future, because some Zoomer-kid used the wrong sushi colour in his code base ......

Emojis in strings are all fine (I still hate them, but that is completely beside the point), however, imagine trying tracking down a bug for hours on end only because your field named :santaclaus:  has the wrong gender or skin colour. IMO, C# code (meaning symbol names - not string literals) should be restricted to UTF-16.

**Reply by Zabadam** [reply]
I see your point, but I counter that if you are using that many similar yet different emoji, you may already have easily-confused UTF-16 variable names in your programming dialect (like missing an 's' because this is an Items[] and not an Item).

**Comment by rockstardev** [comment]
As the Rockstar Developer, I completely support this Proposal. Truth to be told I am shocked that those working on C# language were so restrictive to only allow subset of Unicode characters in variable names. What's next, we'll only have English `chars` with occasional number and symbol like `_`? I don't want to live in an oppressive world where this type of discrimination reigns supreme.

Thus, I urge contributors to this repository to recognize the self-evident truth that all Unicode characters are created equal. And that it's not up to those who designed C# to use it as polygon for enforcing their prejudices. If I want to declare variable using my trusty 🎸 emoji... I should be allowed to do that. It's all between me and my emojis... and teammates - who'll be thrilled that we have more options when it comes to deciding variable names.

🚀

**Comment by cartermp** [comment]
Whole heartedly requesting this for C# 9

As does @blowdart

**Comment by ufcpp** [comment]
What exactly do you want to use as identifiers? spec needed.

FYI:

- [Rust](https://doc.rust-lang.org/reference/identifiers.html): ASCII only
- [Go](https://golang.org/ref/spec#Identifiers): Letters and digits only
- [Java](https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.8): Similar to C#, basically follows [UAX #31](https://unicode.org/reports/tr31/)
- [Swift](https://docs.swift.org/swift-book/ReferenceManual/zzSummaryOfTheGrammar.html): All characters outside [BMP](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane) are identifier-character. This includes Not-Assigned characters, [Variation Selecters](https://en.wikipedia.org/wiki/Variation_Selectors_Supplement), [Musical Symbols](https://en.wikipedia.org/wiki/Musical_Symbols_(Unicode_block)), and [Arrows](https://en.wikipedia.org/wiki/Supplemental_Arrows-C)
- [Emoji data](https://www.unicode.org/Public/UCD/latest/ucd/emoji/emoji-data.txt): If you want to use only Emoji instead of all non-BMP characters, [`IsIdentifierStartCharacter` and `IsIdentifierPartCharacter`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/InternalUtilities/UnicodeCharacterUtilities.cs) need additional data table which is not as simple as [general category](https://docs.microsoft.com/ja-jp/dotnet/api/system.char.getunicodecategory). Moreover, [`#` (U+0023)](https://emojipedia.org/number-sign/) and [`*` (U+002A)](https://emojipedia.org/asterisk/) are included in Emoji.
- [Emoji EBNF and Regex](http://unicode.org/reports/tr51/#EBNF_and_Regex): Some Emoji consist of multiple codepoints. For instance, 👩🏼‍👨🏽‍👦🏼‍👧🏽 consist of 1F469 1F3FC 200D 1F468 1F3FD 200D 1F466 1F3FC 200D 1F467 1F3FD (encoded 41 bytes in UTF-8). Some sequence include BMP characters (e.g. 👨‍⚕️ = 1F468 200D 2695 FE0F) and need more data table.
  - In [C# identifier spec](https://docs.microsoft.com/ja-jp/dotnet/csharp/language-reference/language-specification/lexical-structure), any formatting characters are removed and [ZWJ](https://en.wikipedia.org/wiki/Zero-width_joiner) (U+200D) is a formatting character. 👩🏼‍👨🏽‍👦🏼‍👧🏽 and 👩🏼👨🏽👦🏼👧🏽 (ZWJ removed from  👩🏼‍👨🏽‍👦🏼‍👧🏽 ) are identical in C#.
  - Subdivision flags such as [Flag: Scotland](https://emojipedia.org/flag-scotland/) consist of 🏴 + formatting [Tag characters](https://en.wikipedia.org/wiki/Tags_(Unicode_block)). Thus all subdivision flags are identical to 🏴 ([Black Flag](https://emojipedia.org/emoji/%F0%9F%8F%B4/))

**Comment by sylveon** [comment]
FWIW C++ supports them: https://godbolt.org/z/P37427

The [list of allowed characters in the standard](http://eel.is/c++draft/lex.name#1) seems to allows ZWJ as well

**Comment by CyrusNajmabadi** [comment]
i would champion an effort to get C# to support any/all identifiers that match the unicode identifier annex.  If emojis are part of that annex, then that would naturally fall out.  If they are not, i wouldn't support or do work to make them part of the language for identifiers.

**Comment by ufcpp** [comment]
Emoji is not. Most of Emoji are Other Symbol.

**Comment by CyrusNajmabadi** [comment]
Good to know.  Not championing this issue then.  Thanks!

**Comment by NicolasDorier** [comment]
Apparently Clang 10 support it.

**Comment by NicolasDorier** [comment]
@ufcpp I think the same as C++ would be perfect. http://eel.is/c++draft/lex.name#1

**Comment by ufcpp** [comment]
@NicolasDorier 
Is it actually C++ standard? I found [an article that proposes C++ should follow UAX#31 in C++ 23](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1949r1.html).

**Comment by sylveon** [comment]
eel.is/c++draft is the latest available draft of the C++ standard

**Comment by ufcpp** [comment]
So [the proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1949r1.html) is breaking change? Anyway C++ seems likely to change its  lexical conventions.

**Comment by ufcpp** [comment]
Of course, UAX#31 is not perfect. Some characters are still indistinguishable.

https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGUCWBbADgGwgAUEB7AcwSm2QBoYQE4A7AHwAEAGAAnYEYAbLz4A6AMKlmyUoQDcAWABQSgPQruUbgGpuANSgJMsTJO7oIhAMYxSCbgAoAqloBiAUU49U3d5xcBKJUxmGA1uAF5uACYFRWDQqEAAP4juAGZY+I1AQD+UgFYMkI1AID+UgHYChMBgP5S+PgqNQBA/mvSgwqhAUD+a8taEwDA/moBOeqhAcD+UqJa4tsAIP/Ghno1ASD+U1LqFqEAoP5XuqYTAaD+UgBY13Y1AGD+jycyoQFg/vKu2wDg/vPnTqEB4P5SBE6UAdUMYBAADLBCD2TQAKg0yShUBysJKsOqsKasM6sP6sLGsNmsOWsK2sIOsIusLusOesPe/liQA

**Comment by sylveon** [comment]
Yes, P1949 would be breaking for those who used emojis in code.

**Comment by jfbastien** [comment]
Yes the C++ change is technically breaking, but doesn’t seem to be in practice. We’re likely to adopt a change along those in the paper for C++23. The current discussion is tracked here:
https://github.com/sg16-unicode/sg16/issues/56
Tagging @tahonermann to follow up.

**Comment by tahonermann** [comment]
Just here to elaborate on what @jfbastien has already stated.  We are in the process of vetting a proposal targeted at C++23 that would restrict the set of characters allowed in C++ identifiers.  The most recent published paper is [P1949R3](https://wg21.link/p1949r3).

Since C++11, the set of allowed identifiers as specified in [[lex.name]p1](http://eel.is/c++draft/lex.name#1) corresponds to those proposed and adopted by [N3146](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2010/n3146.html).  That proposal followed the ["Alternative Identifier Syntax" as defined by the Unicode 5.2 version of UAX #31](http://www.unicode.org/reports/tr31/tr31-11.html#Alternative_Identifier_Syntax).  (The "Alternative Identifier Syntax" was later renamed to "Immutable Identifiers" in Unicode 9.0.)

Motivation for this change includes, but is not limited to the following:
- The current set of allowed identifiers permits distinct identifiers that are visually indistinguishable.  This is a security and usability concern since it impacts local reasoning about code; a variable reference might appear to be to a local declaration, but actually resolve against a declaration at file or namespace scope.  This is particularly true when non-normalized identifiers are used.
- The table of allowed identifier code points in [[lex.name]p1](http://eel.is/c++draft/lex.name#1) has not been maintained since Unicode 5.2.
- Recent versions of [UAX #31](https://unicode.org/reports/tr31) note that, if "Immutable identifiers" are used, then the mechanisms described in [UAX #39](https://unicode.org/reports/tr39) cannot be used.
- C++ has no need for immutable identifiers; we can expand the set of allowed identifiers as Unicode adopts new characters.

**Comment by felipeklafke** [comment]
We are not prepared for this evolution. 🙃

**Comment by chucker** [comment]
>Swift: All characters outside BMP are identifier-character. This includes Not-Assigned characters, Variation Selecters, Musical Symbols, and Arrows

Now before you scoff at Musical Symbols, consider that we can't write C♯ the proper way and make it an identifier!

**Comment by Enderlook** [comment]
I don't want to be pessimistic, but, in which serious program would you write emoji member names?
Emoji names are more difficult to read. Not all text editor does support emojis (quote needed). Also, people with sight issues may not distinguish correctly the colour or shape of an emoji.
For example, :blush: and :relaxed: looks almost identical. It's very easy to confuse them.

**Comment by viniciusvillas** [comment]
> I don't want to be pessimistic, but, in which serious program would you write emoji member names?
> Emoji names are more difficult to read. Not all text editor does support emojis (quote needed). Also, people with sight issues may not distinguish correctly the colour or shape of an emoji.
> For example, 😊 and ☺️ looks almost identical. It's very easy to confuse them.

I don't get this necessity either. But I'm kinda old, so go figure what they're trying to accomplish these days. I suspect this is not a work-related issue. :P

**Comment by Unknown6656** [comment]
@viniciusvillas If it were a work-related issue I would switch away from computer science and start studying business astrology :P

**Comment by kurema** [comment]
> I don't want to be pessimistic, but, in which serious program would you write emoji member names?

Typical situation is source-code generation.

For example, XAML code``<Button x:Name="button_😊" Content="😊" />`` should be compiled into ``private global::Windows.UI.Xaml.Controls.Button button_😊;``.
Imagine you are designing something like,
![image](https://user-images.githubusercontent.com/10556974/83088223-15f05300-a0ce-11ea-9f5a-d94c12bee3e2.png).
Names like ``button_SmilingFaceWithSmilingEyes`` or ``button_1F60A`` are very confusing.

Or when you have XSD file like this,
```XML Shema
<xsd:element name="faces">
  <xsd:simpleType>
    <xsd:restriction base="xsd:string">
      <xsd:enumeration value="😊" />
      <xsd:enumeration value="🥺" />
      <xsd:enumeration value="😂" />
      <xsd:enumeration value="😍" />
    </xsd:restriction>
  </xsd:simpleType>
</xsd:element>
```
what result do you expect from xsd.exe other than something like this.
```
public enum faces
{
    😊,
    🥺,
    😂,
    😍
}
```

**Comment by kurema** [comment]
If emoji will not be allowed in member name, we should at least decide how source-code generator should handle emoji.
Then source-code generator programmer know what they should do and IntelliSense can handle it.

**Comment by NickAcPT** [comment]
> > I don't want to be pessimistic, but, in which serious program would you write emoji member names?
> 
> Typical situation is source-code generation.
> 
> For example, XAML code`<Button x:Name="button_😊" Content="😊" />` should be compiled into `private global::Windows.UI.Xaml.Controls.Button button_😊;`.
> Imagine you are designing something like,
> ![image](https://user-images.githubusercontent.com/10556974/83088223-15f05300-a0ce-11ea-9f5a-d94c12bee3e2.png).
> Names like `button_SmilingFaceWithSmilingEyes` or `button_1F60A` are very confusing.
> 
> Or when you have XSD file like this,
> 
> ```
> <xsd:element name="faces">
>   <xsd:simpleType>
>     <xsd:restriction base="xsd:string">
>       <xsd:enumeration value="😊" />
>       <xsd:enumeration value="🥺" />
>       <xsd:enumeration value="😂" />
>       <xsd:enumeration value="😍" />
>     </xsd:restriction>
>   </xsd:simpleType>
> </xsd:element>
> ```
> 
> what result do you expect from xsd.exe other than something like this.
> 
> ```
> public enum faces
> {
>     😊,
>     🥺,
>     😂,
>     😍
> }
> ```

I'm not a big developer or anything like that. Don't work on a company either. But I don't think people actually want to do that. GitHub probably doesn't do it that way either.

I just can't imagine taking a random library from NuGet and trying to access a property or a method, but not being able to do it without having to use code completion because inserting emotes is a bit weird.

**Comment by kurema** [comment]
Well, yes. ItemsControl and ViewModels or Tag or something smarter is the better way in this very case. But what I mean is that allowing emoji is often useful when you are using XAML.
I want ``button_😊_Click`` to be suggested not ``button__Click`` or ``button_1F60A_Click``.

**Comment by Enderlook** [comment]
> > > I don't want to be pessimistic, but, in which serious program would you write emoji member names?
> > 
> > 
> > Typical situation is source-code generation.
> > [...]
> > ```
> > public enum faces
> > {
> >     😊,
> >     🥺,
> >     😂,
> >     😍
> > }
> > ```
> 
> I'm not a big developer or anything like that. Don't work on a company either. But I don't think people actually want to do that. GitHub probably doesn't do it that way either.
> 
> I just can't imagine taking a random library from NuGet and trying to access a property or a method, but not being able to do it without having to use code completion because inserting emotes is a bit weird.

I agree with @NickAcPT. That example sounds as a **very** specific issue @kurema, and I'm quite sure that people don't do that. As @NickAcPT pointed, without IntelliSense, it would become very tedious to program.

<sub>As him, I also don't have experience on the matter.</sub>

**Comment by CyrusNajmabadi** [comment]
i dont' see why my variables would ever have these names.  i.e. if i have an arbitrary emoji picker to tag items with, then it's not like i'm going to have a variable for each possible option.

**Comment by CyrusNajmabadi** [comment]
> If emoji will not be allowed in member name, we should at least decide how source-code generator should handle emoji.

I don't know what this means.  why would source generators have to handle emoji ever?

> Then source-code generator programmer know what they should do and IntelliSense can handle it.

can you clarify what you're discussing here?

**Comment by kurema** [comment]
Situation I imagined is like this.

* Adding thumbs-up and some other emoji button (Not so many).
* You don't have good name for other emoji.
* You want to do emoji specific operation.
* You named Button "Button_(emoji)" because you have no other idea.

Is this that rare situation?
Yes, it's not easy to write without IntelliSense or ClipBoard, but nobody write XAML without VisualStudio.
Allowing emoji in emoji related UI can make things easy. You don't have to think about name of emoji.
At least if there are some rules, even unofficial one, IntelliSense may handle it or you know how to name.

**Comment by CyrusNajmabadi** [comment]
> Is this that rare situation? Yes

In that case, it's definitely not worth doing.   

> You don't have good name for other emoji.

How about `var thumbsUpButton`?

Seems completely understandable.

**Comment by kurema** [comment]
OK. Unicode character name works for many cases.

**Comment by qrli** [comment]
I don't know what you sees, I'm seeing blank squares. Even though unicode covers all characters/symbols, most fonts do not.
![image](https://user-images.githubusercontent.com/10579882/83103438-45608900-a0e9-11ea-84e8-50f002976dde.png)

**Comment by kurema** [comment]
![image](https://user-images.githubusercontent.com/10556974/83104609-0edc3b80-a0f4-11ea-9755-65405fd903e5.png)
 Just random emoji from [emojipedia](https://emojipedia.org/people/).

**Comment by miloush** [comment]
Following UAX #31 sounds like a reasonable and a valid policy. I would be slightly worried about departing from an established standard, regardless of whether it is emoji related or not.

All arguments that we have seen so far could be applied to UAX #31, there is nothing C# specific, so I would suggest trying to raise these requests with UTC. If you manage to convince the Unicode standard to change, then you have a way easier case for requesting the C# language to catch up.

**Comment by lukesdm** [comment]
The year is 2030. The world is run by megacorps. A typical job interview opens with, "Implement a garbage collector using only the following variables: 👶,👨,👴,👻"

**Comment by MgSam** [comment]
I assumed this suggestion was a joke. I'm surprised how much serious discussion has taken place.

**Comment by 333fred** [comment]
Alright everyone. I'd ask that we keep any further discussion to actual serious comments and not more joking. While some might consider it all in good fun and I don't think anyone in this thread has been specifically targeting any individual, this was a legitimate proposal made in good faith and I'd like to keep this thread focused on it. I'll be going through and marking some comments as off topic for the thread.

**Comment by MgSam** [comment]
> Alright everyone. I'd ask that we keep any further discussion to actual serious comments and not more joking. While some might consider it all in good fun and I don't think anyone in this thread has been specifically targeting any individual, this was a legitimate proposal made in good faith and I'd like to keep this thread focused on it. I'll be going through and marking some comments as off topic for the thread.

I'm not convinced that `public enum ThreeStateBool { 🍍, Yes, No }` was made in good faith... ¯\_(ツ)_/¯

**Comment by NicolasDorier** [comment]
I admit this was initially for fun, but the discussion is actually interesting.
C++ supports it only until recently (clang 10.0 and  g++ 10) so I think this is relevant.

**Comment by Serentty** [comment]
@Unknown6656 

> IMO, C# code (meaning symbol names - not string literals) should be restricted to UTF-16.

I don't know what you mean by this. UTF-16 can encode emoji, or else C# strings wouldn't be able to hold them. UTF-16 uses a single 16-bit number for BMP characters, and two 16-bit numbers for characters outside the BMP. If you mean that you think identifiers should be restricted to the BMP, then I strongly disagree. The BMP/non-BMP distinction is purely historical and has nothing to do with suitability for identifiers or how “wacky” a character is. There are multiple living languages that are written in characters partially or entirely outside of the BMP.

As for whether emoji should be allowed in variable names, I think the Unicode recommendations for identifiers are very sound. There's a reason that $ isn't allowed in identifiers, and emoji fall into a similar logical category. The Unicode recommendations are very logical and consistent. For example, they consider numbers in other numeral systems to work along the same logical principles as the Arabic numerals 0 through 9 when it comes to where they can occur in identifiers. I wouldn't agree with further deviating from a standard which has had a lot of effort and thought put into it.

**Comment by NicolasDorier** [comment]
> . There's a reason that $ isn't allowed in identifiers, and emoji fall into a similar logical category. The Unicode recommendations are very logical and consistent.

Where is the reasoning?

**Comment by Serentty** [comment]
The way I see it, the reasoning is this: languages usually don't allow “symbols” in identifiers because they usually use them as part of the syntax of the language instead. Think of all the operators in C# for example. So within ASCII or EBCDIC or whatever old encoding with a small set of characters, generally programming languages allow the set of characters which can be used to write words to be used in identifiers, and reserve other characters for use by language syntax without the need for whitespace delineation. Often numbers are considered okay, as long as they're not at the start of an identifier so there's no confusion with literals.

Now, when we move beyond these old encodings into Unicode, what do we do? The approach that the standard that C# follows takes is essentially just to extend that pattern out, and try to be as consistent as possible. Characters which are used to write words are allowed, just as before. Characters which are used as symbols, punctuation, whitespace, or control are not. Characters which are used only in the writing of numbers in other (human) languages are allowed, but not at the beginning of an identifier. It has a logical consistency to it that you lose if you just say: “This character is really cute, therefore I want to name a class with it.”

If you want fun names, options are still open without emoji. You can name your variables in Ancient Egyptian written in hieroglyphs if you want. Or at least you will be able to once the issues around non-BMP identifiers are fixed in Roslyn.

**Comment by chrisaut** [comment]
Personally I don't care for this. I'm not against it being in the language if it were somehow "free" in terms of cost to get it to work, although I would certainly ban any use of emoji in any code base my team works on.

But it's not free to get it to work, I think the team's time can be spend on much more valuable additions.

**Comment by viniciusvillas** [comment]
> Personally I don't care for this. I'm not against it being in the language if it were somehow "free" in terms of cost to get it to work, although I would certainly ban any use of emoji in any code base my team works on.
> 
> But it's not free to get it to work, I think the team's time can be spend on much more valuable additions.

I second that!

**Comment by CyrusNajmabadi** [comment]
Even if free, I would reject this proposal. So far I haven't seen any compelling cases where this would be valuable or appropriate, and I've seen a lot of cases that show why I would not want this in the language.

**Reply by mcattle** [reply]
> So far I haven't seen any compelling cases where this would be valuable or appropriate, and I've seen a lot of cases that show why I would not want this in the language.

`catch (Exception 💩)`

**Reply by jmarolf** [reply]
You can already do this

```csharp
catch (Exception ಠ_ಠ)
```

So I dunno about the value emoji bring

**Reply by kyoyama-kazusa** [reply]
> You can already do this
> 
> ```cs
> catch (Exception ಠ_ಠ)
> ```
> 
> So I dunno about the value emoji bring

```c#
var _ = 0;
return ~~(0 ^ _ ^ 0);
```

**Comment by Daynvheur** [comment]
> I already see some [...] programming bugs in the future, because some Zoomer-kid used the wrong sushi colour in his code base ......

That's something already possible today with variable name suffixed by an automatic increment, something like `Button5`+`Button23`+`Button42`+..., that are semantically indistinguishables. Bad naming would still be a pain, emoji variations would only be a step anyone could acclimate with.

_Side note: I didn't find any color variation for the :sushi: sushi emoji, but it would be a nice thing for Intellisense to check for emoji 
 possible misreferences, something like "You used :broken_heart:, did you mean :heart: instead?"._

**Comment by masonwheeler** [comment]
> I admit this was initially for fun, but the discussion is actually interesting.
> C++ supports it only until recently (clang 10.0 and g++ 10) so I think this is relevant.

"Well, C++ did it" is generally a pretty good argument in favor of *not* doing something, particularly when the majority of other languages are not doing the same thing.  A lot of the success of the CLR and the C# language is built on learning from C++'s mistakes rather than repeating them.

**Comment by laicasaane** [comment]
Interesting idea but certainly no. Please think about people who are visually impaired. Explicit and long names are always better.

**Comment by NathanielArmer** [comment]
I support this request because I think it would add a lot of value when trying to teach the langue to beginners, or even as a fun way to introduce new features to existing developers.

It's not necessarily a good practice for production code, but not all code is production code. Expressiveness is an important feature of a language, and Emoji are a powerful way of expressing certain idea's.

**Comment by RenderMichael** [comment]
> It's not necessarily a good practice for production code, but not all code is production code.

If this gets implemented, it would have to be paired with an analyzer and a rule to disallow emojis (preferably disallowed by default). Given those conditions, I think it's not a bad idea, but at that point the implementation cost probably skyrockets (someone with an understanding of implementing Roslyn analyzers could determine exactly how costly it would be).

**Reply by yaakov-h** [reply]
That analyser would be less than a day's work for someone experienced with them. Compared to language changes or compiler changes, analysers are trivial.

**Reply by jnm2** [reply]
If I remember @CyrusNajmabadi's comments about studies they did, it's a day's work for someone who hasn't worked with Roslyn analyzers before and more like minutes or hours for someone who has.

**Comment by shayded-exe** [comment]
I want this exclusively so I can do
```cs
throw new 🤷‍♂️();
```

**Reply by Daynvheur** [reply]
You would need something like:
```cs
using 🤷‍♂️ = System.Exception;
```
Which I do agree would be awesome.

**Comment by ArtVandelay101** [comment]
`catch (Exception ☠️)`

**Reply by KennethHoff** [reply]
```c#
catch (☠️)
```

**Reply by ArtVandelay101** [reply]
This proposal is just for variable names. I think you are looking for this one https://www.emojicode.org 😃

**Reply by KennethHoff** [reply]
The only example he gave is not a variable though 😭

**Reply by Daynvheur** [reply]
To be in par with https://github.com/dotnet/csharplang/discussions/3490#discussioncomment-8253748:
```cs
using ☠️ = System.Exception;
//...
catch(☠️); // catch(System.Exception);
```

**Comment by Kerrick** [comment]
C++ isn't the only language that supports emoji identifiers. You can also do it in Ruby.

```
kerrick@💻👨🏻☝🏻 ~ % irb
irb(main):001> RUBY_VERSION
=> "3.4.2"
irb(main):002> 🧙 = "Elanduilinilithar of house Bronsuiligifinabinen"
=> "Elanduilinilithar of house Bronsuiligifinabinen"
irb(main):003> puts 🧙
Elanduilinilithar of house Bronsuiligifinabinen
=> nil
```

**Reply by AndersKehlet** [reply]
I need this for interop with F#

```fsharp
type ``🤔``<'T> = Option<'T>
```

Well, I say "need", but it's more of a deep yearning that I don't quite understand.
